pkg/controller/keycloakuser/keycloakuser_reconciler.go

```diff
@@ -0,0 +1,208 @@
+package keycloakuser
+
+import (
+	"fmt"
+
+	"github.com/keycloak/keycloak-operator/pkg/apis/keycloak/v1alpha1"
+	"github.com/keycloak/keycloak-operator/pkg/common"
+)
+
+type Reconciler interface {
+	Reconcile(cr *v1alpha1.KeycloakUser) error
+}
+
+type KeycloakuserReconciler struct {
+	Realm string
+}
+
+func NewKeycloakuserReconciler(realm string) *KeycloakuserReconciler {
+	return &KeycloakuserReconciler{
+		Realm: realm,
+	}
+}
+
+func (i *KeycloakuserReconciler) Reconcile(state *common.UserState, cr *v1alpha1.KeycloakUser) common.DesiredClusterState {
+	if cr.DeletionTimestamp != nil {
+		return i.reconcileUserDelete(state, cr)
+	}
+	return i.reconcileUser(state, cr)
+}
+
+func (i *KeycloakuserReconciler) reconcileUser(state *common.UserState, cr *v1alpha1.KeycloakUser) common.DesiredClusterState {
+	desired := common.DesiredClusterState{}
+
+	desired.AddAction(i.getKeycloakDesiredState())
+	desired.AddActions(i.getKeycloakUserDesiredState(state, cr))
+
+	return desired
+}
+
+func (i *KeycloakuserReconciler) reconcileUserDelete(state *common.UserState, cr *v1alpha1.KeycloakUser) common.DesiredClusterState {
+	desired := common.DesiredClusterState{}
+
+	desired.AddAction(i.getKeycloakDesiredState())
```

Comment:
Obviously, this code won't scale...

I guess it's fine for now but once you have a free slot - please shoot an email to Keycloak Dev mailing list and describe this problem. It needs to by solved in Keycloak itself.

Reply:
This first action only pings Keycloak to make sure it's available for the client. Plus this is only ran when a user CR is updated/deleted.